lib/openstack_query/api/query_api.py

```diff
@@ -285,3 +285,28 @@ def append_from(
 
         self.output.update_forwarded_outputs(link_props[0], new_query.to_props())
         return self
+
+    def reset(self, hard=False):
+        """
+        function that when called will reset the query back to original state.
+        This means that any pre-configured info like select(), where(), group_by() and sort_by() will be deleted
+        :param hard: boolean (default False). If true will delete any cached results from previous run() command
+        """
+        # reset select
+        self.output.selected_props = set()
+
+        # reset where
+        self.builder.client_side_filters = []
+        self.builder.server_side_filters = []
+        self.builder.server_filter_fallback = []
+
+        # reset group_by
+        self.parser.group_by = None
+        self.parser.group_mappings = {}
+
+        # reset sort_by
+        self.parser.sort_by = {}
+
+        if hard:
+            self.executer.raw_results = []
+        return self
```

Comment:
This means that `query_api` now knows how the internal of a different class works. Can we not use the existing factory to regenerate the builder / parser / etc... since it means a reset is the same as an init?

Reply:
alternatively I could implement a reset() on each block that the API method calls

Reply:
I don't want the Query API accessing the Factory - all kinds of circular import headaches